File: backend/src/agentic_traveler/orchestrator/profile_questions.py
"""Traveler-DNA question bank — Python mirror of ``frontend/src/lib/profile-questions.ts`` (Task 54).

Single source of truth for the questions the just-in-time elicitation engine
(Task 55) may weave into chat to build the "living" Traveler DNA. The TS registry
is the canonical author surface; this module is a hand-kept mirror (the monorepo
isolation rule means the backend cannot import a frontend file at runtime — the
Cloud Run image ships ``backend/`` only). ``tests/orchestrator/test_profile_questions.py``
guards the two against drift, mirroring the existing ``CAPABILITY_INTENTS`` sync test.

``binding`` is the key axis:
  - ``profile``    — a stable trait. Answered once, persisted to
                     ``user_profiles.profile_data.answered_questions``, reused.
  - ``flow_state`` — a current-state question (the "different I's"): re-asked each
                     time the user (re)enters the flow. Never persisted durably.

This is the SEED set; the nuanced expansion (~30–45 questions) is authored in Task 59.
"""

from __future__ import annotations

from dataclasses import dataclass

PROFILE = "profile"
FLOW_STATE = "flow_state"


@dataclass(frozen=True)
class ProfileChoiceDef:
    id: str
    label: str
    value: str


@dataclass(frozen=True)
class ProfileQuestionDef:
    id: str
    binding: str  # "profile" | "flow_state"
    prompt: str
    choices: tuple[ProfileChoiceDef, ...]
    informs: tuple[str, ...]
    category: str  # "compass" | "pulse" | "strategy" | "identity" | "state"
    cost: str  # "tap" | "flash_lite"
    allow_multi: bool = False
    tally_key: str | None = None
    hard_override_slot: str | None = None


def _c(id: str, label: str, value: str) -> ProfileChoiceDef:
    return ProfileChoiceDef(id=id, label=label, value=value)


PROFILE_QUESTIONS: tuple[ProfileQuestionDef, ...] = (
    # ── profile · compass ──────────────────────────────────────────────────
    ProfileQuestionDef(
        id="travel_company",
        binding=PROFILE,
        prompt="Who's usually in your travel bubble?",
        choices=(
            _c("solo", "Solo — just me", "solo"),
            _c("duo", "My partner", "duo"),
            _c("inner_circle", "Close friends or family", "inner_circle"),
            _c("socialite", "I meet people on the road", "socialite"),
        ),
        informs=("social_energy", "travel_company"),
        category="compass",
        cost="tap",
        tally_key="travel_bubble",
    ),
    ProfileQuestionDef(
        id="trip_success",
        binding=PROFILE,
        prompt="What makes a trip feel like a success?",
        choices=(
            _c("nature", "Nature", "nature"),
            _c("culture", "Culture & history", "culture"),
            _c("taste", "Food & markets", "taste"),
            _c("peace", "Peace & reset", "peace"),
            _c("adventure", "Adventure", "adventure"),
            _c("aesthetics", "Beautiful places", "aesthetics"),
            _c("social", "People & nightlife", "social"),
            _c("growth", "Learning & growth", "growth"),
        ),
        informs=("themes", "motivations"),
        category="compass",
        cost="tap",
        allow_multi=True,
        tally_key="trip_success",
    ),
    ProfileQuestionDef(
        id="meaning_depth",
        binding=PROFILE,
        prompt="How much does the deeper meaning of a place matter to you?",
        choices=(
            _c("seeker", "I seek meaning & connection", "seeker"),
            _c("explorer", "I enjoy it, but I'm here for the experience", "explorer"),
            _c("vacationer", "I care more about the vibe & fun", "vacationer"),
        ),
        informs=("meaning_depth", "spiritual_interest"),
        category="compass",
        cost="tap",
        tally_key="meaning_depth",
    ),
    ProfileQuestionDef(
        id="immersion",
        binding=PROFILE,
        prompt="Where's your sweet spot for local immersion?",
        choices=(
            _c("deep_end", "Eat & stay like a local", "deep_end"),
            _c("best_of_both", "Authentic, but a comfy bed", "best_of_both"),
            _c("curated", "Curated & high-end", "curated"),
        ),
        informs=("immersion", "comfort_preference"),
        category="compass",
        cost="tap",
        tally_key="immersion",
    ),
    # ── profile · pulse ────────────────────────────────────────────────────
    ProfileQuestionDef(
        id="morning_vibe",
        binding=PROFILE,
        prompt="What's your ideal morning on a trip?",
        choices=(
            _c("sunrise", "Up for the sunrise", "sunrise"),
            _c("civilized", "A civilized, slow start", "civilized"),
            _c("night_owl", "Mornings are for sleeping", "night_owl"),
        ),
        informs=("daily_rhythm",),
        category="pulse",
        cost="tap",
        tally_key="ideal_morning",
    ),
    ProfileQuestionDef(
        id="activity_intensity",
        binding=PROFILE,
        prompt="How much sweat equity are you up for?",
        choices=(
            _c("leisurely", "Leisurely", "leisurely"),
            _c("active", "Active", "active"),
            _c("rugged", "Rugged", "rugged"),
        ),
        informs=("activity_intensity",),
        category="pulse",
        cost="tap",
        tally_key="physical_intensity",
    ),
    ProfileQuestionDef(
        id="pace",
        binding=PROFILE,
        prompt="What's your pace when you travel?",
        choices=(
            _c("sprinter", "See it all (fast)", "fast"),
            _c("wanderer", "A bit of both (medium)", "medium"),
            _c("slow_mo", "Slow & deep", "slow"),
        ),
        informs=("pace", "energy_strategy"),
        category="pulse",
        cost="tap",
        tally_key="energy_strategy",
    ),
    # ── profile · strategy ─────────────────────────────────────────────────
    ProfileQuestionDef(
        id="structure_preference",
        binding=PROFILE,
        prompt="How much do you like a plan vs. wandering?",
        choices=(
            _c("total_freedom", "Total freedom", "total_freedom"),
            _c("flexible", "A loose skeleton", "flexible"),
            _c("planned", "Mostly planned", "planned"),
        ),
        informs=("structure_preference", "spontaneity"),
        category="strategy",
        cost="tap",
        tally_key="solo_freedom",
    ),
    ProfileQuestionDef(
        id="risk_appetite",
        binding=PROFILE,
        prompt="How much discomfort are you happy to embrace?",
        choices=(
            _c("comfort_first", "Comfort is the priority", "comfort_first"),
            _c("balanced", "A bit of friction is fine", "balanced"),
            _c("discomfort_ok", "Discomfort is part of the story", "discomfort_ok"),
        ),
        informs=("risk_appetite",),
        category="strategy",
        cost="tap",
        tally_key="uncertainty_scale",
    ),
    ProfileQuestionDef(
        id="adaptability",
        binding=PROFILE,
        prompt="A strike cancels your transport. What's your move?",
        choices=(
            _c("tactician", "Find the workaround", "tactician"),
            _c("alchemist", "Make it the new plan", "alchemist"),
            _c("delegate", "Let me handle it for you", "delegate"),
        ),
        informs=("adaptability",),
        category="strategy",
        cost="tap",
        tally_key="strike_scenario",
    ),
    ProfileQuestionDef(
        id="splurge_priority",
        binding=PROFILE,
        prompt="Where does the treat-yourself money usually go?",
        choices=(
            _c("stay", "The stay", "stay"),
            _c("plate", "The food", "plate"),
            _c("experience", "The experience", "experience"),
            _c("piece", "Local craft or art", "piece"),
        ),
        informs=("splurge_priority",),
        category="strategy",
        cost="tap",
        tally_key="splurge",
    ),
    ProfileQuestionDef(
        id="budget_tier",
        binding=PROFILE,
        prompt="What's your budget personality?",
        choices=(
            _c("negotiator", "The negotiator", "negotiator"),
            _c("balanced", "Balanced", "balanced"),
            _c("high_end", "High-end", "high_end"),
        ),
        informs=("budget_tier",),
        category="strategy",
        cost="tap",
        tally_key="budget_personality",
        hard_override_slot="ask.budget",
    ),
    ProfileQuestionDef(
        id="deal_breakers",
        binding=PROFILE,
        prompt="Any absolute deal-breakers?",
        choices=(
            _c("poor_hygiene", "Poor hygiene", "poor_hygiene"),
            _c("no_wifi", "No Wi-Fi", "no_wifi"),
            _c("crowds", "Overwhelming crowds", "crowds"),
            _c("extreme_heat", "Extreme heat", "extreme_heat"),
            _c("extreme_cold", "Extreme cold", "extreme_cold"),
            _c("poor_transport", "Poor transport", "poor_transport"),
        ),
        informs=("avoids",),
        category="strategy",
        cost="tap",
        allow_multi=True,
        tally_key="deal_breakers",
    ),
    # ── flow_state · the "different I's" (re-asked each flow run) ───────────
    ProfileQuestionDef(
        id="trip_intent_this_time",
        binding=FLOW_STATE,
        prompt="What are you hoping this trip gives you?",
        choices=(
            _c("reset", "Rest & reset", "reset"),
            _c("adventure", "Adventure", "adventure"),
            _c("connection", "Connection", "connection"),
            _c("discovery", "Discovery", "discovery"),
            _c("everything", "A bit of everything", "everything"),
        ),
        informs=("trip_intent",),
        category="state",
        cost="tap",
    ),
    ProfileQuestionDef(
        id="energy_for_this_trip",
        binding=FLOW_STATE,
        prompt="How's your energy for this one?",
        choices=(
            _c("low", "Low — keep it gentle", "low"),
            _c("steady", "Steady", "steady"),
            _c("high", "High — let's go", "high"),
        ),
        informs=("energy_state",),
        category="state",
        cost="tap",
    ),
    ProfileQuestionDef(
        id="current_craving",
        binding=FLOW_STATE,
        prompt="Right now, what sounds best?",
        choices=(
            _c("nature", "Nature", "nature"),
            _c("culture", "Culture", "culture"),
            _c("food", "Food", "food"),
            _c("nightlife", "Nightlife", "nightlife"),
            _c("quiet", "Quiet", "quiet"),
        ),
        informs=("craving",),
        category="state",
        cost="tap",
    ),
)

BY_ID: dict[str, ProfileQuestionDef] = {q.id: q for q in PROFILE_QUESTIONS}


def legal_option_values(qid: str) -> set[str]:
    """Legal ``value`` set for a question — used to re-validate a selection
    server-side (trust-but-verify; the client registry is never trusted)."""
    q = BY_ID.get(qid)
    return {c.value for c in q.choices} if q else set()

File: frontend/src/lib/profile-questions.ts
// profile-questions.ts — the single Traveler-DNA question bank (Task 54).
//
// One source of truth for every question the app may ask to build the "living"
// Traveler DNA. Consumed by the just-in-time elicitation engine (Task 55, a Python
// mirror in backend/.../orchestrator/profile_questions.py kept in sync) and rendered
// through the existing tappable-chip mechanism (SlotRequest.choices, Task 43).
//
// Pure data + types — no React, no lucide imports — so it can be consumed anywhere.
//
// `binding` is the key axis:
//   profile    — a stable trait. Answered once, persisted to
//                user_profiles.profile_data.answered_questions, reused across sagas.
//   flow_state — a current-state question (the "different I's"): re-asked each time
//                the user (re)enters the flow. Never persisted to the durable profile.
//
// This is the SEED set. The nuanced expansion (target ~30–45 questions) is authored
// in Task 59. The seed intentionally includes the small flow_state trio that the
// sagas wire in Tasks 55/56 (trip_intent_this_time, energy_for_this_trip,
// current_craving) so those references resolve before Task 59 lands.

export type QuestionBinding = "profile" | "flow_state";
export type QuestionCategory =
  | "compass"
  | "pulse"
  | "strategy"
  | "identity"
  | "state";

export type ProfileChoice = { id: string; label: string; value: string };

export type ProfileQuestion = {
  id: string; // stable snake_case, e.g. "travel_company"
  binding: QuestionBinding;
  prompt: string; // <= 200 chars, one question
  choices: ProfileChoice[]; // tappable options; [] => free-text only
  allowMulti?: boolean; // e.g. "what makes a trip feel successful"
  informs: string[]; // DNA dimensions / tag families this answer feeds
  category: QuestionCategory;
  cost: "tap" | "flash_lite"; // tap = deterministic write; flash_lite = text parse
  tallyKey?: string; // maps to a legacy Tally form_response key, for backfill (Task 54 §7 Step 8)
  hardOverrideSlot?: string; // if this hard_override slot is set, the question is "covered"
};

export const PROFILE_QUESTIONS: ProfileQuestion[] = [
  // ── profile · compass ──────────────────────────────────────────────────────
  {
    id: "travel_company",
    binding: "profile",
    prompt: "Who's usually in your travel bubble?",
    choices: [
      { id: "solo", label: "Solo — just me", value: "solo" },
      { id: "duo", label: "My partner", value: "duo" },
      { id: "inner_circle", label: "Close friends or family", value: "inner_circle" },
      { id: "socialite", label: "I meet people on the road", value: "socialite" },
    ],
    informs: ["social_energy", "travel_company"],
    category: "compass",
    cost: "tap",
    tallyKey: "travel_bubble",
  },
  {
    id: "trip_success",
    binding: "profile",
    prompt: "What makes a trip feel like a success?",
    choices: [
      { id: "nature", label: "Nature", value: "nature" },
      { id: "culture", label: "Culture & history", value: "culture" },
      { id: "taste", label: "Food & markets", value: "taste" },
      { id: "peace", label: "Peace & reset", value: "peace" },
      { id: "adventure", label: "Adventure", value: "adventure" },
      { id: "aesthetics", label: "Beautiful places", value: "aesthetics" },
      { id: "social", label: "People & nightlife", value: "social" },
      { id: "growth", label: "Learning & growth", value: "growth" },
    ],
    allowMulti: true,
    informs: ["themes", "motivations"],
    category: "compass",
    cost: "tap",
    tallyKey: "trip_success",
  },
  {
    id: "meaning_depth",
    binding: "profile",
    prompt: "How much does the deeper meaning of a place matter to you?",
    choices: [
      { id: "seeker", label: "I seek meaning & connection", value: "seeker" },
      { id: "explorer", label: "I enjoy it, but I'm here for the experience", value: "explorer" },
      { id: "vacationer", label: "I care more about the vibe & fun", value: "vacationer" },
    ],
    informs: ["meaning_depth", "spiritual_interest"],
    category: "compass",
    cost: "tap",
    tallyKey: "meaning_depth",
  },
  {
    id: "immersion",
    binding: "profile",
    prompt: "Where's your sweet spot for local immersion?",
    choices: [
      { id: "deep_end", label: "Eat & stay like a local", value: "deep_end" },
      { id: "best_of_both", label: "Authentic, but a comfy bed", value: "best_of_both" },
      { id: "curated", label: "Curated & high-end", value: "curated" },
    ],
    informs: ["immersion", "comfort_preference"],
    category: "compass",
    cost: "tap",
    tallyKey: "immersion",
  },

  // ── profile · pulse ────────────────────────────────────────────────────────
  {
    id: "morning_vibe",
    binding: "profile",
    prompt: "What's your ideal morning on a trip?",
    choices: [
      { id: "sunrise", label: "Up for the sunrise", value: "sunrise" },
      { id: "civilized", label: "A civilized, slow start", value: "civilized" },
      { id: "night_owl", label: "Mornings are for sleeping", value: "night_owl" },
    ],
    informs: ["daily_rhythm"],
    category: "pulse",
    cost: "tap",
    tallyKey: "ideal_morning",
  },
  {
    id: "activity_intensity",
    binding: "profile",
    prompt: "How much sweat equity are you up for?",
    choices: [
      { id: "leisurely", label: "Leisurely", value: "leisurely" },
      { id: "active", label: "Active", value: "active" },
      { id: "rugged", label: "Rugged", value: "rugged" },
    ],
    informs: ["activity_intensity"],
    category: "pulse",
    cost: "tap",
    tallyKey: "physical_intensity",
  },
  {
    id: "pace",
    binding: "profile",
    prompt: "What's your pace when you travel?",
    choices: [
      { id: "sprinter", label: "See it all (fast)", value: "fast" },
      { id: "wanderer", label: "A bit of both (medium)", value: "medium" },
      { id: "slow_mo", label: "Slow & deep", value: "slow" },
    ],
    informs: ["pace", "energy_strategy"],
    category: "pulse",
    cost: "tap",
    tallyKey: "energy_strategy",
  },

  // ── profile · strategy ─────────────────────────────────────────────────────
  {
    id: "structure_preference",
    binding: "profile",
    prompt: "How much do you like a plan vs. wandering?",
    choices: [
      { id: "total_freedom", label: "Total freedom", value: "total_freedom" },
      { id: "flexible", label: "A loose skeleton", value: "flexible" },
      { id: "planned", label: "Mostly planned", value: "planned" },
    ],
    informs: ["structure_preference", "spontaneity"],
    category: "strategy",
    cost: "tap",
    tallyKey: "solo_freedom",
  },
  {
    id: "risk_appetite",
    binding: "profile",
    prompt: "How much discomfort are you happy to embrace?",
    choices: [
      { id: "comfort_first", label: "Comfort is the priority", value: "comfort_first" },
      { id: "balanced", label: "A bit of friction is fine", value: "balanced" },
      { id: "discomfort_ok", label: "Discomfort is part of the story", value: "discomfort_ok" },
    ],
    informs: ["risk_appetite"],
    category: "strategy",
    cost: "tap",
    tallyKey: "uncertainty_scale",
  },
  {
    id: "adaptability",
    binding: "profile",
    prompt: "A strike cancels your transport. What's your move?",
    choices: [
      { id: "tactician", label: "Find the workaround", value: "tactician" },
      { id: "alchemist", label: "Make it the new plan", value: "alchemist" },
      { id: "delegate", label: "Let me handle it for you", value: "delegate" },
    ],
    informs: ["adaptability"],
    category: "strategy",
    cost: "tap",
    tallyKey: "strike_scenario",
  },
  {
    id: "splurge_priority",
    binding: "profile",
    prompt: "Where does the treat-yourself money usually go?",
    choices: [
      { id: "stay", label: "The stay", value: "stay" },
      { id: "plate", label: "The food", value: "plate" },
      { id: "experience", label: "The experience", value: "experience" },
      { id: "piece", label: "Local craft or art", value: "piece" },
    ],
    informs: ["splurge_priority"],
    category: "strategy",
    cost: "tap",
    tallyKey: "splurge",
  },
  {
    id: "budget_tier",
    binding: "profile",
    prompt: "What's your budget personality?",
    choices: [
      { id: "negotiator", label: "The negotiator", value: "negotiator" },
      { id: "balanced", label: "Balanced", value: "balanced" },
      { id: "high_end", label: "High-end", value: "high_end" },
    ],
    informs: ["budget_tier"],
    category: "strategy",
    cost: "tap",
    tallyKey: "budget_personality",
    hardOverrideSlot: "ask.budget",
  },
  {
    id: "deal_breakers",
    binding: "profile",
    prompt: "Any absolute deal-breakers?",
    choices: [
      { id: "poor_hygiene", label: "Poor hygiene", value: "poor_hygiene" },
      { id: "no_wifi", label: "No Wi-Fi", value: "no_wifi" },
      { id: "crowds", label: "Overwhelming crowds", value: "crowds" },
      { id: "extreme_heat", label: "Extreme heat", value: "extreme_heat" },
      { id: "extreme_cold", label: "Extreme cold", value: "extreme_cold" },
      { id: "poor_transport", label: "Poor transport", value: "poor_transport" },
    ],
    allowMulti: true,
    informs: ["avoids"],
    category: "strategy",
    cost: "tap",
    tallyKey: "deal_breakers",
  },

  // ── flow_state · the "different I's" (re-asked each flow run) ───────────────
  {
    id: "trip_intent_this_time",
    binding: "flow_state",
    prompt: "What are you hoping this trip gives you?",
    choices: [
      { id: "reset", label: "Rest & reset", value: "reset" },
      { id: "adventure", label: "Adventure", value: "adventure" },
      { id: "connection", label: "Connection", value: "connection" },
      { id: "discovery", label: "Discovery", value: "discovery" },
      { id: "everything", label: "A bit of everything", value: "everything" },
    ],
    informs: ["trip_intent"],
    category: "state",
    cost: "tap",
  },
  {
    id: "energy_for_this_trip",
    binding: "flow_state",
    prompt: "How's your energy for this one?",
    choices: [
      { id: "low", label: "Low — keep it gentle", value: "low" },
      { id: "steady", label: "Steady", value: "steady" },
      { id: "high", label: "High — let's go", value: "high" },
    ],
    informs: ["energy_state"],
    category: "state",
    cost: "tap",
  },
  {
    id: "current_craving",
    binding: "flow_state",
    prompt: "Right now, what sounds best?",
    choices: [
      { id: "nature", label: "Nature", value: "nature" },
      { id: "culture", label: "Culture", value: "culture" },
      { id: "food", label: "Food", value: "food" },
      { id: "nightlife", label: "Nightlife", value: "nightlife" },
      { id: "quiet", label: "Quiet", value: "quiet" },
    ],
    informs: ["craving"],
    category: "state",
    cost: "tap",
  },
];

export function profileQuestionById(id: string): ProfileQuestion | undefined {
  return PROFILE_QUESTIONS.find((q) => q.id === id);
}

/** Legal option values for a question — used to re-validate a selection. */
export function legalOptionValues(id: string): Set<string> {
  const q = profileQuestionById(id);
  return new Set((q?.choices ?? []).map((c) => c.value));
}
